File: examples/cpp/src/host_capture_calibration.cpp
#include <qnbot/glove.hpp>

#include <algorithm>
#include <chrono>
#include <iostream>
#include <optional>
#include <stdexcept>
#include <string>
#include <utility>
#include <vector>

namespace {

constexpr const char* operator_id = "default";
constexpr const char* left_source = "primary-glove-left";
constexpr const char* right_source = "primary-glove-right";
constexpr const char* target_name = "openxr_hand";
constexpr const char* builtin_skeleton_package_id =
    "qnbot_hand.dynamic_openxr_hand";
constexpr auto wait_timeout = std::chrono::minutes(5);

struct Options {
    std::vector<std::string> package_ids;
    bool force{false};
};

struct ConfiguredTarget {
    std::string package_id;
    std::string name;
};

Options parse_options(int argc, char** argv) {
    Options options;
    for (int index = 1; index < argc; ++index) {
        const std::string argument = argv[index];
        if (argument == "--package-id") {
            if (++index >= argc)
                throw std::invalid_argument("--package-id requires a value");
            const std::string package_id = argv[index];
            if (std::find(options.package_ids.begin(),
                          options.package_ids.end(),
                          package_id) == options.package_ids.end()) {
                options.package_ids.push_back(package_id);
            }
        } else if (argument == "--force") {
            options.force = true;
        } else {
            throw std::invalid_argument("unknown argument: " + argument);
        }
    }
    if (options.package_ids.empty())
        throw std::invalid_argument("--package-id is required");
    return options;
}

std::vector<ConfiguredTarget>
configured_targets(const std::vector<std::string>& package_ids) {
    const auto ordinary_count = static_cast<std::size_t>(std::count_if(
        package_ids.begin(), package_ids.end(), [](const std::string& value) {
            return value != builtin_skeleton_package_id;
        }));
    std::size_t ordinary_index = 0;
    std::vector<ConfiguredTarget> targets;
    for (const auto& package_id : package_ids) {
        if (package_id == builtin_skeleton_package_id) {
            targets.push_back({package_id, "skeleton"});
            continue;
        }
        ++ordinary_index;
        targets.push_back(
            {package_id, ordinary_count == 1
                             ? target_name
                             : std::string(target_name) + "_" +
                                   std::to_string(ordinary_index)});
    }
    return targets;
}

qnbot::SdkConfig make_config(const std::vector<std::string>& package_ids) {
    qnbot::GloveConfig left;
    left.name = left_source;
    left.side = qnbot::Side::left;
    qnbot::GloveConfig right;
    right.name = right_source;
    right.side = qnbot::Side::right;

    qnbot::SdkConfig config;
    config.devices = {left, right};
    for (const auto& target : configured_targets(package_ids)) {
        if (target.package_id == builtin_skeleton_package_id) continue;
        for (const auto& side_and_source :
             {std::make_pair(qnbot::Side::left, left_source),
              std::make_pair(qnbot::Side::right, right_source)}) {
            qnbot::TargetConfig configured;
            configured.name = target.name;
            configured.side = side_and_source.first;
            configured.source = qnbot::DeviceSelector{
                "glove", std::string(side_and_source.second), std::nullopt};
            configured.algorithms = {qnbot::TargetAlgorithm{target.package_id}};
            config.targets.push_back(std::move(configured));
        }
    }
    return config;
}

std::string select_target(const std::vector<std::string>& package_ids) {
    const auto targets = configured_targets(package_ids);
    if (targets.size() == 1) return targets.front().name;
    for (std::size_t index = 0; index < targets.size(); ++index) {
        std::cout << index + 1 << ": " << targets[index].name << " ("
                  << targets[index].package_id << ")\n";
    }
    while (true) {
        std::cout << "target [1-" << targets.size() << "]: ";
        std::string answer;
        std::getline(std::cin, answer);
        try {
            const auto selected = static_cast<std::size_t>(std::stoul(answer));
            if (selected >= 1 && selected <= targets.size())
                return targets[selected - 1].name;
        } catch (const std::exception&) {
        }
        std::cout << "enter one of the listed target numbers\n";
    }
}

void update_or_wait(qnbot::Glove& glove) {
    const auto update = glove.update();
    if (update.has_next_task()) static_cast<void>(update.sleep());
}

qnbot::CaptureSession capture_side(qnbot::Glove& glove,
                                   const std::vector<std::string>& package_ids,
                                   const std::string& source, qnbot::Side side,
                                   bool force, bool start_runtime) {
    qnbot::CaptureStatusRequest status_request;
    status_request.package_ids = package_ids;
    status_request.operator_id = operator_id;
    status_request.source = source;
    status_request.side = side;
    const auto readiness = glove.capture_status(status_request);
    std::cout << (side == qnbot::Side::left ? "left" : "right")
              << ": capture readiness=" << static_cast<int>(readiness.state)
              << '\n';
    for (const auto& stage : readiness.stages)
        std::cout << "  " << stage.stage_id << '\n';

    qnbot::CaptureStartRequest request;
    request.package_ids = package_ids;
    request.operator_id = operator_id;
    request.source = source;
    request.side = side;
    request.force = force;
    auto session = glove.start_capture(request);
    if (start_runtime) glove.start();
    const auto deadline = std::chrono::steady_clock::now() + wait_timeout;
    while (std::chrono::steady_clock::now() < deadline) {
        const auto snapshot = session.snapshot();
        if (snapshot.state == qnbot::CaptureSessionState::completed) {
            if (!session.result())
                throw std::runtime_error(
                    "completed capture session has no result");
            return session;
        }
        if (snapshot.state == qnbot::CaptureSessionState::failed ||
            snapshot.state == qnbot::CaptureSessionState::cancelled) {
            throw std::runtime_error(
                snapshot.failure ? snapshot.failure->message
                                 : std::string("capture session stopped"));
        }

        const auto current = std::find_if(
            snapshot.stage_progress.begin(), snapshot.stage_progress.end(),
            [&](const qnbot::CaptureStageProgress& stage) {
                return snapshot.current_stage &&
                       stage.stage_id == *snapshot.current_stage;
            });
        if (current != snapshot.stage_progress.end() &&
            current->state == qnbot::CaptureStageState::awaiting_confirmation) {
            if (!snapshot.request_id) {
                session.cancel();
                throw std::runtime_error(
                    "capture confirmation is missing its request token");
            }
            const auto plan_stage = std::find_if(
                snapshot.plan.stages.begin(), snapshot.plan.stages.end(),
                [&](const qnbot::CapturePlanStage& stage) {
                    return snapshot.current_stage &&
                           stage.stage_id == *snapshot.current_stage;
                });
            std::cout << (plan_stage == snapshot.plan.stages.end()
                              ? "Continue capture"
                              : plan_stage->prompt)
                      << " [Y/n]: ";
            std::string answer;
            std::getline(std::cin, answer);
            if (answer.empty() || answer == "confirm" || answer == "y" ||
                answer == "yes") {
                session.confirm(*snapshot.request_id);
            } else {
                session.cancel();
                throw std::runtime_error("capture cancelled by operator");
            }
            continue;
        }
        update_or_wait(glove);
    }
    session.cancel();
    throw std::runtime_error("capture timed out");
}

void calibrate(qnbot::Glove& glove,
               const qnbot::CaptureSession& capture_session,
               const std::vector<std::optional<std::string>>& targets) {
    auto calibration = glove.start_calibration(capture_session, targets);
    const auto deadline = std::chrono::steady_clock::now() + wait_timeout;
    while (std::chrono::steady_clock::now() < deadline) {
        const auto snapshot = calibration.snapshot();
        if (snapshot.state == qnbot::CalibrationSessionState::completed) {
            const auto results = calibration.result();
            if (!results) {
                calibration.close();
                throw std::runtime_error(
                    "completed calibration session has no result");
            }
            for (const auto& result : *results)
                std::cout << "saved calibration record for " << result.target_id
                          << '\n';
            calibration.close();
            return;
        }
        const auto failed = std::find_if(
            snapshot.jobs.begin(), snapshot.jobs.end(),
            [](const qnbot::CalibrationJobProgress& job) {
                return job.state == qnbot::CalibrationJobState::failed;
            });
        if (failed != snapshot.jobs.end()) {
            const std::string message = failed->failure
                                            ? failed->failure->message
                                            : "calibration failed";
            calibration.close();
            throw std::runtime_error(message);
        }
        if (snapshot.state == qnbot::CalibrationSessionState::cancelled) {
            calibration.close();
            throw std::runtime_error("calibration cancelled");
        }
        update_or_wait(glove);
    }
    calibration.close();
    throw std::runtime_error("calibration timed out");
}

} // namespace

int main(int argc, char** argv) {
    try {
        const auto options = parse_options(argc, argv);
        qnbot::Sdk sdk(make_config(options.package_ids));
        auto glove = sdk.glove();
        if (std::find(options.package_ids.begin(), options.package_ids.end(),
                      builtin_skeleton_package_id) !=
            options.package_ids.end()) {
            static_cast<void>(glove.device(left_source).skeleton());
            static_cast<void>(glove.device(right_source).skeleton());
        }

        auto left_session =
            capture_side(glove, options.package_ids, left_source,
                         qnbot::Side::left, options.force, true);
        auto right_session =
            capture_side(glove, options.package_ids, right_source,
                         qnbot::Side::right, options.force, false);
        const std::vector<std::optional<std::string>> targets{
            select_target(left_session.snapshot().plan.package_ids)};
        calibrate(glove, left_session, targets);
        calibrate(glove, right_session, targets);

        right_session.close();
        left_session.close();
        glove.close();
        std::cout << "Calibration results are saved and applied to retargeting "
                     "by the SDK.\n";
        return 0;
    } catch (const std::exception& error) {
        std::cerr << "host capture and calibration failed: " << error.what()
                  << '\n';
        return 1;
    }
}

File: examples/python/src/host_capture_calibration.py
from __future__ import annotations

import argparse

from qnbot_sdk import DeviceSelector, Sdk, Side, TargetConfig, TargetType
from qnbot_sdk.glove import (
    CalibrationJobState,
    CaptureControlAction,
    CaptureReadiness,
    CaptureSessionState,
    CaptureSessionSnapshot,
    CaptureSet,
    CaptureStageState,
    Glove,
    GloveConfig,
)

OPERATOR_ID = "default"
LEFT_SOURCE = "primary-glove-left"
RIGHT_SOURCE = "primary-glove-right"
TARGET_NAME = "openxr_hand"


def create_sdk() -> Sdk:
    return Sdk(
        devices=(
            GloveConfig(name=LEFT_SOURCE, side=Side.LEFT),
            GloveConfig(name=RIGHT_SOURCE, side=Side.RIGHT),
        ),
        targets=(
            TargetConfig(
                type=TargetType.HAND,
                name=TARGET_NAME,
                side=Side.LEFT,
                source=DeviceSelector(type="glove", name=LEFT_SOURCE),
                algorithms=(),
            ),
            TargetConfig(
                type=TargetType.HAND,
                name=TARGET_NAME,
                side=Side.RIGHT,
                source=DeviceSelector(type="glove", name=RIGHT_SOURCE),
                algorithms=(),
            ),
        ),
    )


def render_capture_readiness(side: Side, readiness: CaptureReadiness) -> None:
    resolved = ", ".join(
        f"{package.id}@{package.version}"
        for package in readiness.plan.resolved_packages
    )
    print(f"{side.value}: capture readiness={readiness.state.value} plan={resolved}")
    readiness_by_stage = {stage.stage_id: stage for stage in readiness.stages}
    for stage in readiness.plan.stages:
        current = readiness_by_stage[stage.stage_id]
        requirement = "optional" if stage.optional else "required"
        print(
            f"  {stage.stage_id}: {current.state.value}, {requirement}, "
            f"samples={stage.required_sample_count}, prompt={stage.prompt}"
        )


def render_session_snapshot(side: Side, snapshot: CaptureSessionSnapshot) -> None:
    operations = ",".join(action.value for action in snapshot.allowed_operations)
    print(
        f"{side.value}: session={snapshot.state.value} "
        f"current={snapshot.current_stage or '-'} "
        f"progress={snapshot.collected_sample_count}/"
        f"{snapshot.required_sample_count} next={snapshot.next_action} "
        f"allowed={operations or 'none'}"
    )
    for stage in snapshot.stage_runs:
        print(
            f"  {stage.stage_id}: state={stage.state.value}, "
            f"origin={stage.origin.value}, "
            f"samples={stage.collected_sample_count}/"
            f"{stage.required_sample_count}"
        )


def render_capture_set(side: Side, capture_set: CaptureSet) -> None:
    snapshot = capture_set.snapshot()
    print(f"{side.value}: completed CaptureSet source={snapshot.source_id}")
    for stage in snapshot.stage_samples:
        print(f"  {stage.stage_id}: saved samples={stage.frame_count}")


def capture_side(
    glove: Glove,
    candidate_package_ids: tuple[str, ...],
    source: str,
    side: Side,
    force: bool,
) -> CaptureSet:
    readiness = glove.capture_status(
        candidate_ids=candidate_package_ids,
        operator_id=OPERATOR_ID,
        source=source,
        side=side,
    )
    render_capture_readiness(side, readiness)
    session = glove.start_capture(
        candidate_ids=candidate_package_ids,
        operator_id=OPERATOR_ID,
        source=source,
        side=side,
        force=force,
    )
    try:
        shown_snapshot: CaptureSessionSnapshot | None = None
        while True:
            snapshot = session.snapshot()
            if snapshot != shown_snapshot:
                render_session_snapshot(side, snapshot)
                shown_snapshot = snapshot

            if snapshot.state is CaptureSessionState.COMPLETED:
                result = session.result()
                if result is None:
                    raise RuntimeError("completed capture session has no result")
                render_capture_set(side, result)
                return result
            if snapshot.state in (
                CaptureSessionState.FAILED,
                CaptureSessionState.CANCELLED,
            ):
                raise RuntimeError(snapshot.failure or f"{side.value} capture stopped")
            if snapshot.state is CaptureSessionState.AWAITING_CONFIRMATION:
                if snapshot.request_id is None:
                    raise RuntimeError(
                        "capture confirmation is missing its request token"
                    )
                current_stage = snapshot.current_stage or "current stage"
                prompt = next(
                    (
                        stage.prompt
                        for stage in snapshot.plan.stages
                        if stage.stage_id == snapshot.current_stage
                    ),
                    current_stage,
                )
                allowed = set(snapshot.allowed_operations)
                retryable_stage_ids = tuple(
                    stage.stage_id
                    for stage in snapshot.stage_runs
                    if stage.state
                    in (CaptureStageState.CAPTURED, CaptureStageState.REUSED)
                )
                choices = []
                if CaptureControlAction.CONFIRM in allowed:
                    choices.append("Enter=confirm")
                if CaptureControlAction.SKIP in allowed:
                    choices.append("skip=skip")
                if CaptureControlAction.RETRY in allowed:
                    choices.append(
                        f"retry <stage_id> ({', '.join(retryable_stage_ids)})"
                    )
                if CaptureControlAction.CANCEL in allowed:
                    choices.append("cancel=cancel")
                answer = (
                    input(f"{side.value} {prompt} [{', '.join(choices)}]: ")
                    .strip()
                    .lower()
                )
                action, _, retry_stage_id = answer.partition(" ")
                if answer == "cancel" and CaptureControlAction.CANCEL in allowed:
                    session.cancel()
                elif answer == "skip" and CaptureControlAction.SKIP in allowed:
                    session.skip(snapshot.request_id)
                elif (
                    action == "retry"
                    and CaptureControlAction.RETRY in allowed
                    and retry_stage_id in retryable_stage_ids
                ):
                    session.retry(retry_stage_id)
                elif (
                    answer in ("", "confirm", "y", "yes")
                    and CaptureControlAction.CONFIRM in allowed
                ):
                    session.confirm(snapshot.request_id)
                else:
                    print("action is not allowed in the current session state")
                continue

            update = glove.update()
            if update.has_next_task:
                update.sleep()
    finally:
        session.close()


def choose_package(argument: str | None, candidate_package_ids: tuple[str, ...]) -> str:
    selected = (
        argument
        or input(f"Choose package ({', '.join(candidate_package_ids)}): ").strip()
    )
    if selected not in candidate_package_ids:
        raise ValueError("selected package must be one of the captured candidates")
    return selected


def calibrate(glove: Glove, capture_set: CaptureSet, package_id: str) -> None:
    job = glove.start_calibration(
        capture_set=capture_set,
        package_id=package_id,
        target=TARGET_NAME,
    )
    try:
        shown_state: CalibrationJobState | None = None
        while True:
            snapshot = job.snapshot()
            if snapshot.state is not shown_state:
                print(
                    f"calibration state={snapshot.state.value} "
                    f"package={snapshot.package_id}@{snapshot.package_version}"
                )
                shown_state = snapshot.state
            if snapshot.state is CalibrationJobState.COMPLETED:
                result = job.result()
                if result is None:
                    raise RuntimeError("completed calibration job has no result")
                print(
                    f"saved {result.package_id}@{result.package_version} "
                    f"for {result.target_id}"
                )
                return
            if snapshot.state is CalibrationJobState.FAILED:
                failure = snapshot.failure
                raise RuntimeError(
                    failure.message if failure is not None else "calibration failed"
                )
            update = glove.update()
            if update.has_next_task:
                update.sleep()
    finally:
        job.close()


def main() -> None:
    arguments = argparse.ArgumentParser(
        description=(
            "Inspect reuse, capture both hands for a candidate package set, "
            "then calibrate the selected package"
        )
    )
    arguments.add_argument(
        "--package-id",
        action="append",
        dest="package_ids",
        required=True,
        metavar="PACKAGE_ID",
        help=(
            "candidate package to include in the merged CapturePlan; repeat for "
            "multiple packages"
        ),
    )
    arguments.add_argument("--force", action="store_true")
    options = arguments.parse_args()
    candidate_package_ids = tuple(options.package_ids)

    sdk = create_sdk()
    glove = sdk.glove()
    left_capture_set: CaptureSet | None = None
    right_capture_set: CaptureSet | None = None
    try:
        glove.connect()
        glove.start()

        left_capture_set = capture_side(
            glove,
            candidate_package_ids,
            LEFT_SOURCE,
            side=Side.LEFT,
            force=options.force,
        )
        right_capture_set = capture_side(
            glove,
            candidate_package_ids,
            RIGHT_SOURCE,
            side=Side.RIGHT,
            force=options.force,
        )

        selected_package_id = (
            candidate_package_ids[0]
            if len(candidate_package_ids) == 1
            else choose_package(None, candidate_package_ids)
        )
        calibrate(glove, left_capture_set, selected_package_id)
        calibrate(glove, right_capture_set, selected_package_id)
        print("Calibration results are saved and applied to retargeting by the SDK.")
    finally:
        if right_capture_set is not None:
            right_capture_set.close()
        if left_capture_set is not None:
            left_capture_set.close()
        glove.close()


if __name__ == "__main__":
    main()
